python/custom_field_definitions/add_custom_field.py

```diff
@@ -0,0 +1,183 @@
+    for cdf in custom_field_defs_resp:
+        custom_field_def = {}
+        custom_field_def['name'] = cdf['name']
+        custom_field_def['id'] = cdf['id']
+        custom_field_def['data_type'] = cdf['data_type']
+        custom_field_def['entity_type'] = cdf['definition_entity_type']
+        custom_field_def['facet'] = cdf['facet']
+
+        # Add custom field definition records to custom field defintions keyed by name.
+        custom_field_defs[cdf['name']] = custom_field_def
+
+    return custom_field_defs
+
+# Compare current custom field defintion obtained via the API, to the custom field defintion
+# in the CSV file.
+def compare_custom_field_defs(curr_cfd, csv_cfd):
+    if curr_cfd['data_type'] != csv_cfd['data_type']:
+        print(f"Data_types do not agree.  {curr_cfd['data_type']} != CVS {csv_cfd['data_type']}")
+        sys.exit(1)
+    if curr_cfd['entity_type'] != csv_cfd['entity_type']:
+        print(f"Entity_types do not agree.  {curr_cfd['entity_type']} != CVS {csv_cfd['entity_type']}")
+        sys.exit(1)
+
+    return curr_cfd['id']
+
+# Process the CSV input file custom field defintion dictionary.
+# Return a dictionary of custom field defintions with custom field defintion parameters,
+# custom field values, and vulnerability ID.
+def process_input_file(csv_input_file_name):
+    cdfs = {}
+
+    try:
+        with open(csv_input_file_name, newline='', mode='r', encoding='utf-8-sig') as input_file:
+            reader = csv.DictReader(input_file, delimiter=',', skipinitialspace=True) 
+            for row in reader:
+                cdf_name = row['name']
+                cdfs[cdf_name] = row
+
+    except FileNotFoundError:
+        print(f"ERROR: CSV input file, {csv_input_file_name} not found.")
+        sys.exit(1)
+
+    return cdfs
+
+# Create a custom field defintion.
+def create_custom_field_definition(base_url, headers, cfd_dict):
+    create_url = f"{base_url}custom_field_definitions"
+
+    payload = {}
+    payload['name'] = cfd_dict['name']
+    payload['description'] = cfd_dict['description']
+    payload['data_type'] = cfd_dict['data_type']
+    payload['definition_entity_type'] = cfd_dict['entity_type']
+    payload['facet'] = cfd_dict['facet']
+    if 'dropdown_options' in cfd_dict:
+        payload['drop_down_options'] = cfd_dict['dropdown_options']
+
+    response = requests.post(create_url, headers=headers, data=json.dumps(payload))
+    resp_json = response.json()
+    if response.status_code != 201:
+        print_http_error("Create Custom Field Definition Error: ", response.status_code, resp_json)
+        sys.exit(1)
+
+    custom_field_def = resp_json['custom_field_definition']
+    return custom_field_def['id']
+
+# Update a custom field in a vulnerability.
+def update_vuln(base_url, headers, vuln_id, custom_field_def_id, custom_field_value):
+    vuln_id = vuln_id.strip()
+
+    update_url = f"{base_url}vulnerabilities/{vuln_id}"
```

Comment:
I am not a big fan of `base_url` containing `/` it would be a nicer read if 
`update_url = f"{base_url}/vulnerabilities/{vuln_id}"`

Reply:
Changed.